internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,679 @@
+	iterProducers := make([]*iteratorProducer, 0, len(children))
+	for idx, child := range children {
+		producerChan, err := c.fastPathRewrite(ctx, req, child)
+		if err != nil {
+			return nil, err
+		}
+		iterProducers = append(iterProducers, &iteratorProducer{idx: idx, producer: producerChan})
+	}
+	var resolver fastPathSetHandler
+	switch op {
+	case unionSetOperator:
+		resolver = fastPathUnion
+	case intersectionSetOperator:
+		resolver = fastPathIntersection
+	case exclusionSetOperator:
+		resolver = fastPathDifference
+	default:
+		return nil, ErrUnknownSetOperator
+	}
+	resultChan := make(chan *iteratorMsg, len(children))
+	go resolver(ctx, iterProducers, resultChan)
+	return resultChan, nil
+}
+
+func (c *LocalChecker) fastPathRewrite(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+	rewrite *openfgav1.Userset,
+) (chan *iteratorMsg, error) {
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgav1.Userset_This:
+		return c.fastPathDirect(ctx, req)
+	case *openfgav1.Userset_ComputedUserset:
+		return c.fastPathComputed(ctx, req, rewrite)
+	case *openfgav1.Userset_Union:
+		return c.fastPathOperationSetup(ctx, req, unionSetOperator, rw.Union.GetChild()...)
+	case *openfgav1.Userset_Intersection:
+		return c.fastPathOperationSetup(ctx, req, intersectionSetOperator, rw.Intersection.GetChild()...)
+	case *openfgav1.Userset_Difference:
+		return c.fastPathOperationSetup(ctx, req, exclusionSetOperator, rw.Difference.GetBase(), rw.Difference.GetSubtract())
+	default:
+		return nil, ErrUnknownSetOperator
+	}
+}
+
+// resolveFastPath attempts to find the intersection across 2 producers (channels) of ObjectIDs.
+// In the case of a TTU:
+// Right channel is the result set of the Read of ObjectID/Relation that yields the User's ObjectID.
+// Left channel is the result set of ReadStartingWithUser of User/Relation that yields Object's ObjectID.
+// From the perspective of the model, the left hand side of a TTU is the computed relationship being expanded.
+func (c *LocalChecker) resolveFastPath(ctx context.Context, chans []chan *iteratorMsg, iter TupleMapper) (*ResolveCheckResponse, error) {
+	rightOpen := true
+	leftOpen := true
+
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	leftChan := fanInIteratorChannels(cancellableCtx, chans)
+	rightChan := streamedLookupUsersetFromIterator(cancellableCtx, iter)
+
+	defer func() {
+		cancel()
+		iter.Stop()
+		if !leftOpen {
+			return
+		}
+		go func() {
+			for msg := range leftChan {
+				if msg.iter != nil {
+					msg.iter.Stop()
+				}
+			}
+		}()
+	}()
+
+	res := &ResolveCheckResponse{
+		Allowed: false,
+	}
+
+	rightSet := hashset.New()
+	leftSet := hashset.New()
+
+	r, ok := <-rightChan
+	if !ok {
+		return res, nil
+	}
+	if r.err != nil {
+		return nil, r.err
+	}
+	rightSet.Add(r.userset)
+	for leftOpen || rightOpen {
```

Comment:
Could this be consolidated to a single value?

Reply:
Elaborate. Keep in mind they are used different contexts.

Reply:
probably can replace with a counter of open channels and this would be `for countopen > 0`, but i like it the way it is 🤷‍♀️